File: node/rabbit-protobuf/src/MessageQueuePublisher.ts
import {Message} from "google-protobuf";

import RabbitMqProvider from "./RabbitMqProvider";


class MessageQueuePublisher {

    // noinspection JSUnusedGlobalSymbols
    public publish<T extends Message>(message: T): void {
        const eventName = (message as any).name;
        RabbitMqProvider.channel.publish(
            RabbitMqProvider.exchangeName,
            eventName,
            Buffer.from(message.serializeBinary()),
            {
                persistent: true,
                deliveryMode: 2
            }
        );

        console.debug(`Emitted ${eventName} event`);
    }

}


export default new MessageQueuePublisher();

File: dotnet/RabbitProtobuf/MessageQueuePublisher.cs
using Google.Protobuf;
using Microsoft.Extensions.Logging;
using RabbitMQ.Client;


namespace RabbitProtobuf {

    internal class MessageQueuePublisher : IMessageQueuePublisher {

        private readonly ILogger<RabbitMqProvider> logger;

        private readonly RabbitMqProvider rabbitMqProvider;



        public MessageQueuePublisher(RabbitMqProvider rabbitMqProvider, ILogger<RabbitMqProvider> logger) {
            this.rabbitMqProvider = rabbitMqProvider;
            this.logger = logger;
        }



        public void Publish(IMessage message) {
            var eventNameStr = message.GetType().FullName;
            rabbitMqProvider.Channel.BasicPublish(
                rabbitMqProvider.ExchangeName,
                eventNameStr,
                rabbitMqProvider.Properties,
                message.ToByteArray()
            );

            logger.LogInformation("Emitted {eventName} event", eventNameStr);
        }

    }

}